jax/scipy/sparse/linalg.py

```diff
@@ -28,25 +28,9 @@
 _vdot = partial(jnp.vdot, precision=lax.Precision.HIGHEST)
 
 
-def _vdot_real_part(x, y):
-  """Vector dot-product guaranteed to have a real valued result."""
-  # all our uses of vdot() in CG are for computing an operator of the form
-  # `z^T M z` where `M` is positive definite and Hermitian, so the result is
-  # real valued:
-  # https://en.wikipedia.org/wiki/Definiteness_of_a_matrix#Definitions_for_complex_matrices
-  result = _vdot(x.real, y.real)
-  if jnp.iscomplexobj(x) or jnp.iscomplexobj(y):
-    result += _vdot(x.imag, y.imag)
-  return result
-
-
 # aliases for working with pytrees
-
-def _vdot_tree(x, y, assume_real=True):
```

Comment:
Does this change affect the CG implementation? I believe it used the `assume_real=True` branch. Would the CG calls to `_vdot_tree` need to be wrapped with `jnp.abs` as well?

Reply:
Well, the tests seem to still pass, but I'll take a look

Reply:
Yes, we definitely want to still use `_vdot_real_part` for CG. Otherwise we are doing a bunch of redundant computation for the imaginary part of the solution, even though we know a priori that it is zero.

Reply:
OK, fair enough. I think the assume_real is a little confusing though, so I made this two different functions.